ACF syntax forward compatibility

george-mcintyre opened the pull request:
### Description

This pull request introduces future-proofing for ACF parsing by allowing files to include unsupported sections without raising errors.  As long as the file is formatted according to the rules it will parse correctly.  
- Allowed and ignored:
  - RULE containing unsupported blocks (RULE ignored)
  - RULE with unsupported permissions (RULE ignored)
  - Top level items named other than supported keywords (may include optional parameter lists, blocks))
  - Sub items in ASG groups named other than supported ASG keywords (may include optional parameter lists, blocks)
  - Sub items in RULE body (may include optional parameter lists, blocks)
- Disallowed with error
  - Sub items in HAG list
  - Sub items in UAG list
  - non standard CALC
  - RULE flags can only be `TRAPWRITE`, or `NOTRAPWRITE`
 
### Checklist

- [x] Ensure backward compatibility.
- [x] Verify functionality with different ACF file examples.

george-mcintyre committed:
Introduce Future Proofing to ACF parsing.

- Allow ACF files to contain sections that are unsupported
- introduce a pragma disable-strict-parsing to disable strict parsing and allow files to be accepted

mdavidsaver commented:
As an example.  The ACF syntax as understood by [`p4p.gw`](https://github.com/epics-base/p4p/blob/master/src/p4p/asLib/yacc.py).

```

asconfig    : asconfig        asconfig_item
asconfig    : asconfig_item

string_list : string_list ',' STRING
asg_list    : asg_list        asg_item
rule_list   : rule_list       rule_item
    
asconfig_item : uag
              | hag
              | asg
rule_item : uag_ref
          | hag_ref
          | calc_head
            
uag : UAG  '(' STRING ')' string_body
hag : HAG  '(' STRING ')' string_body
asg : ASG  '(' STRING ')' asg_body

string_list : STRING
asg_list    : asg_item
rule_list   : rule_item

uag_ref  : UAG  '(' string_list ')'
hag_ref  : HAG  '(' string_list ')'
calc_head : CALC '(' STRING ')'

string_body :
asg_body    :
rule_body   :

string_body : '{' string_list '}'
asg_body    : '{' asg_list    '}'
rule_body   : '{' rule_list   '}'

asg_item : INP '(' STRING ')'
asg_item : RULE '(' INTEGER ',' STRING trap ')' rule_body

trap : ',' STRING
## STRING must be TRAPWRITE or NOTRAPWRITE
```

george-mcintyre committed:
Add tests for future proof parser functionaluty

Expanded tests for Access Security configurations to validate parsing behaviors for unsupported elements. Added detailed cases covering invalid formats, unsupported modifications, and acceptable configurations while maintaining future-proofing and backward compatibility assurances.

george-mcintyre committed:
Refactor parser to handle generic items and floating-point tokens

Replaced "unsupported elements" with a more flexible "generic items" mechanism to enhance parser extensibility. Added support for floating-point numbers in parsing rules. Updated tests and rules to reflect these changes, ensuring better handling of nested and unknown configurations.

george-mcintyre commented:
Updated grammar `asLib.y` and lexer `asLib.l` and added tests `aslibtest.c` to verify functionality.  Follows guidelines specified in [Apr 3, 2025 | EPICS Core F2F meeting - Minutes](https://docs.google.com/document/d/e/2PACX-1vTUZChfGAkvSHNxJolUT6uhRdzHcBr-FtOB3EcbkUfbd1_LaB3V2U9cm4YclanfBFDnfElfo58VHEdw/pub)

The BNF grammar is now as follows:

```bnf
<asconfig> ::= <asconfig> <asconfig_item>
             | <asconfig_item>

/* Top level items with strict keyword handling */
<asconfig_item> ::= "UAG" <uag_head> <uag_body>?
                  | "HAG" <hag_head> <hag_body>?
                  | "ASG" <asg_head> <asg_body>?
                  | <unreserved_name> <generic_head> <generic_block>?

<unreserved_name> ::= STRING except {"UAG", "HAG", "ASG"}

/* UAG definition */
<uag_head> ::= "(" STRING ")"
<uag_body> ::= "{" <uag_user_list> "}"
<uag_user_list> ::= <uag_user_list> "," STRING
                  | STRING

/* HAG definition */
<hag_head> ::= "(" STRING ")"
<hag_body> ::= "{" <hag_host_list> "}"
<hag_host_list> ::= <hag_host_list> "," STRING
                  | STRING

/* ASG definition */
<asg_head> ::= "(" STRING ")"
<asg_body> ::= "{" <asg_body_list> "}"
<asg_body_list> ::= <asg_body_list> <asg_body_item>
                  | <asg_body_item>

<asg_body_item> ::= <inp_config>
                  | <rule_config>
                  | <any_identifier> <generic_head> <generic_block>?

/* Input configuration */
<inp_config> ::= "INP" <link> "(" STRING ")"
<link> ::= "A" | "B" | "C" | "D" | "E" | "F" 
         | "G" | "H" | "I" | "J" | "K" | "L"

/* Rule configuration */
<rule_config> ::= "RULE" <rule_head> <rule_body>?
<rule_head> ::= "(" <rule_head_mandatory> "," <rule_log_option>? ")"
<rule_head_mandatory> ::= INTEGER "," STRING
<rule_log_option> ::= "TRAPWRITE" | "NOTRAPWRITE"

<rule_body> ::= "{" <rule_content_list> "}"
<rule_content_list> ::= <rule_content_list> <rule_content>
                     | <rule_content>

<rule_content> ::= <rule_item>
                 | <any_identifier> <generic_head> <generic_block>?

<rule_item> ::= <uag_ref>
              | <hag_ref>
              | <calc_head>

<uag_ref> ::= "UAG" "(" <string_list> ")"
<hag_ref> ::= "HAG" "(" <string_list> ")"
<calc_head> ::= "CALC" "(" STRING ")"

/* Generic syntax elements */
<any_identifier> ::= STRING

<generic_head> ::= "(" <generic_list>? ")"

<generic_list> ::= <generic_list> "," <generic_element>
                | <generic_element>

<generic_element> ::= STRING
                   | INTEGER
                   | FLOAT

<generic_block> ::= "{" <generic_block_elem_list> "}"

<generic_block_elem_list> ::= <generic_block_elem_list> <generic_block_elem>
                           | <generic_block_elem>
                           | <generic_list>

<generic_block_elem> ::= <any_identifier> <generic_head> <generic_block>?

/* Common elements */
<string_list> ::= <string_list> "," STRING
                | STRING
```

george-mcintyre committed:
Add warnings for unsupported ACF elements and update tests

Introduced `yywarn` to log warnings for unsupported ACF elements, rule permissions, and conditions in `asLib.y`. Updated test configuration to replace generic placeholders with more descriptive names. Adjusted test cases and reduced the test plan count to reflect the changes.

george-mcintyre committed:
Remove unused strictParsing logic and improve error handling.

Removed strictParsing variable and related code as it was no longer in use. Enhanced error messaging and parsing logic for handling unsupported elements. Updated test cases for better clarity and expanded coverage of edge cases.

anjohnson commented on modules/libcom/src/as/asLib_lex.l:
```diff
@@ -48,6 +48,11 @@ INP{link} {
         return(tokenINP);
 }
 
+{digit}*\.{digit}+([eE][-+]?{digit}+)? {
```
Please compare with db file parser, this is missing a leading `[+-]?` pattern (as is the integer below, please fix that too).

anjohnson commented on modules/libcom/src/as/asLib.y:
```diff
@@ -24,14 +25,20 @@ static ASGRULE *yyAsgRule=NULL;
 
 %union
 {
     int Int;
```
Please replace with `epicsInt64`, and adjust other code as necessary, `atoi()` to the `epicsParse..()` in epicsStdlib.h etc.

anjohnson commented on modules/libcom/src/as/asLib.y:
```diff
@@ -123,40 +211,36 @@ inp_config: tokenINP '(' tokenSTRING ')'
+
 
-rule_head_manditory:    '(' tokenINTEGER ',' tokenSTRING
+rule_head_manditory:    tokenINTEGER ',' tokenSTRING
```
Spelled as "mandatory".

anjohnson commented on modules/libcom/src/as/asLib.y:
```diff
@@ -24,14 +25,20 @@ static ASGRULE *yyAsgRule=NULL;
 %token tokenUAG tokenHAG tokenASG tokenRULE tokenCALC
 %token <Str> tokenINP
 %token <Int> tokenINTEGER
+%token <Float> tokenFLOAT
```
Please rename to `tokenINT64` and `tokenDOUBLE`.

george-mcintyre replied:
```
tokenINT64   ~= epicsInt64
tokenFLOAT64 ~= epicsFloat64
```

anjohnson commented on modules/libcom/src/as/asLib.y:
```diff
@@ -213,6 +304,13 @@ static int yyerror(char *str)
+static int yywarn(char *str, char *token)
+{
+    if (!yyWarned && strlen(str) && strlen(token))
+        errlogPrintf("%s at line %d: %s\n", str, line_num, token);
```
Use `printf(stderr, ERL_WARNING ...)` so errors are synchronous and colored, not delayed. Also change `yyerror()` to use `printf`.

george-mcintyre replied:
```
static int yyerror(char *str)
{
    if (strlen(str))
        fprintf(stderr, ERL_ERROR " %s at line %d\n", str, line_num);
    else
        fprintf(stderr, ERL_ERROR " at line %d\n", line_num);
    yyFailed = TRUE;
    return 0;
}
static int yywarn(char *str, char *token)
{
    if (!yyWarned && strlen(str) && strlen(token))
        fprintf(stderr, ERL_WARNING " %s at line %d: %s\n", str, line_num, token);
    yyWarned = TRUE;
    return 0;
}

```

anjohnson commented on modules/libcom/test/aslibtest.c:
```diff
@@ -22,18 +22,527 @@ static char *asUser,
+    "HAG(foo) {localhost}\n"
+
+    "COMPLEX_ARGUMENTS(1, WELL, \"FORMED\", ARG, LIST) {\n"
+    "\tALSO_GENERIC(WELL, FORMED, ARG, LIST, 2.0) \n"
```
Add -ve numbers.

george-mcintyre replied:
```
static const char supported_config_5[] = ""
    "HAG(foo) {localhost}\n"

    "RECURSIVE_SUB_BLOCKS(1.0, -2.3, +4.5, ARGS, +2.71828E-23, -2.71828e+23, +12, -13, +-14) {\n"
    "    ALSO_GENERIC() {\n"
    "        AND_RECURSIVE(FOO) {\n"
    "            LIST, BODY\n"
    "        }\n"
    "    }\n"
    "}\n"

    "ASG(DEFAULT) {\n"
    "    RULE(0, NONE)\n"
    "}\n"

    "ASG(ro) {\n"
    "    RULE(0, NONE)\n"
    "    RULE(+1, READ) {\n"
    "        HAG(foo)\n"
    "    }\n"
    "}\n";

```

anjohnson commented on modules/libcom/test/aslibtest.c:
```diff
@@ -22,18 +22,527 @@ static char *asUser,
+    "\tASG(HAL, IMP, CALC)\n"
+    "\tHAL(USG, MAL) {\n"
+    "\t\tHAG(foo)\n"
+    "\t}\n"
```
`s/\t/  /` perhaps?

anjohnson commented on modules/libcom/test/aslibtest.c:
```diff
@@ -22,18 +22,527 @@ static char *asUser,
+    "\t\tBAD_PREDICATE_AS_WELL(\"EPICS Certificate Authority\")\n"
+    "\t}\n"
+    "}\n";
+
```
Add case where an unknown top-level element is defined and used in a `RULE` body.

george-mcintyre replied:
```
static const char supported_config_6[] = ""
    "HAG(foo) {localhost}\n"

    "WITH_KEYWORDS(UAG) {\n"
    "    ASG(HAL, IMP, CALC, RULE)\n"
    "    HAL(USG, MAL) {\n"
    "        HAG(foo)\n"
    "    }\n"
    "}\n"

    "ASG(DEFAULT) {\n"
    "    RULE(0, NONE)\n"
    "}\n"

    "ASG(ignored) {\n"
    "    RULE(0, NONE)\n"
    "    RULE(1, READ) {\n"
    "        WITH_KEYWORDS(UAG)\n"
    "    }\n"
    "}\n"

    "ASG(ro) {\n"
    "    RULE(0, NONE)\n"
    "    RULE(1, READ) {\n"
    "        HAG(foo)\n"
    "    }\n"
    "    RULE(2, WRITE) {\n"
    "        WITH_KEYWORDS(UAG)\n"
    "    }\n"
    "}\n";

```

anjohnson reviewed (changes requested):
Core Group: Thanks!

mdavidsaver commented on modules/libcom/src/as/asLib.y:
```diff
@@ -43,6 +50,87 @@ asconfig_item:  tokenUAG uag_head uag_body
     |   tokenHAG hag_head
     |   tokenASG asg_head asg_body
     |   tokenASG asg_head
+    |   generic_item
+    ;
+
+keyword: tokenUAG
+    | tokenHAG
+    | tokenASG
+    | tokenRULE
+    | tokenCALC
+    | tokenINP
```
```
2025-04-23T14:55:03.9867849Z antelope: warning - line 57 of "../as/asLib.y", the default action assigns an undefined value to $$
2025-04-23T14:55:03.9868781Z antelope: warning - line 58 of "../as/asLib.y", the default action assigns an undefined value to $$
2025-04-23T14:55:03.9869322Z antelope: warning - line 59 of "../as/asLib.y", the default action assigns an undefined value to $$
2025-04-23T14:55:03.9869846Z antelope: warning - line 60 of "../as/asLib.y", the default action assigns an undefined value to $$
2025-04-23T14:55:03.9870559Z antelope: warning - line 61 of "../as/asLib.y", the default action assigns an undefined value to $$
2025-04-23T14:55:03.9870948Z antelope: 4 shift/reduce conflicts.
2025-04-23T14:55:03.9871240Z "../../../../bin/linux-x86_64/antelope" -basLib  ../as/asLib.y
```

george-mcintyre committed:
Fix typo in 'rule_head_mandatory' definition in asLib.y

george-mcintyre committed:
Enhance ASLib parsing with Int64/Float64 support.

Updated ASLib to utilize 64-bit integers and floats for improved parsing precision. Replaced legacy parsing methods with epics-friendly functions, improved error handling, and adjusted tests and grammar accordingly.

george-mcintyre committed:
Refactor AS parser to separate non-rule keywords

Introduced a new `non_rule_keyword` type and moved `tokenASG` and `tokenRULE` under it for better logical grouping. This improves parser clarity and maintains a clean separation of keyword types.

george-mcintyre committed:
Add rule-specific parsing for generic block elements

Introduce new rules for parsing rule-specific generic block elements and their names. This enhances the grammar to handle additional cases while maintaining memory management and error handling. Adjust references to use the new rule where applicable.

george-mcintyre committed:
Refactor grammar rules in asLib.y to remove remaining shift/reduce conflict

Reorganize and refine grammar rules for handling generic elements, blocks, and lists. Simplifies the structure while introducing clearer distinctions between different constructs. This enhances maintainability and removes ambiguities in parsing.

george-mcintyre commented:
All requested changes applied

anjohnson commented:
Core: @dirk-zimoch, @mdavidsaver please try breaking this before our next meeting.

anjohnson commented:
Core: Needs Release Notes documenting the new generic ACF syntax.
Will also need adding to the RTD chapter on [Access Security](https://docs.epics-controls.org/en/latest/appdevguide/AccessSecurity.html).

dirk-zimoch commented:
> Core: @dirk-zimoch, @mdavidsaver please try breaking this before our next meeting.

Builds without problems for all of my architectures and loads any of our existing asg config files without error and they work as expected.
Ignores RULE containing unknown keywords instead of bailing out with syntax error, as documented.

Accepted and ignored:
```
RULE(1,WRITE) { UNKNOWN(whatever)  }
RULE(1,WRITE) { UNKNOWN(whatever) { something } } # no inner ()
RULE(1,WRITE) { UNKNOWN() }  # empty ()
RULE(1,WRITE) { UNKNOWN() { something } }
RULE(1,WRITE) { UNKNOWN() { something() { xxx } } }
RULE(1,WRITE) { UNKNOWN() { x() { y() { z() } } } }
RULE(1,WRITE) { UNKNOWN() { x() { y() { "string" } } } }
RULE(1,WRITE) { UNKNOWN() { x() { y() { "}" } } } }
RULE(1,WRITE) { UNKNOWN() { x() { y() { "}", something, else } } } }
```
Rejected:
```
RULE(1,WRITE) { UNKNOWN }  # missing ()
RULE(1,WRITE) { UNKNOWN(whatever) {} } # empty {}
RULE(1,WRITE) { UNKNOWN() { something { xxx } } }  # missing () but { ... }
RULE(1,WRITE) { UNKNOWN() { something(whatever) { } } } # empty { }
RULE(1,WRITE) { UNKNOWN() { something else } }  # missing ,
```

Otherwise finds the same syntax errors as before.

Looks fine to me.

dirk-zimoch commented on modules/libcom/src/as/asLib.y:
```diff
@@ -123,40 +239,40 @@ inp_config: tokenINP '(' tokenSTRING ')'
+    {
+        if ($1 < 0) {
+            char message[40];
+            sprintf(message, "RULE: LEVEL must be positive: %lld", $1);
```
Possible `message` buffer overrun when level has more than 10 digits, e.g. `-1000000000000000000`.

dirk-zimoch replied:
Maybe check for level > 1 as well? We currently have only level 0 and 1.

mdavidsaver replied:
Maybe?

```c
char message[51];
epicsSnprintf(message, sizeof(message), "RULE: LEVEL must be positive: %lld", $1);
message[sizeof(message)-1] = '\0';
```

mdavidsaver replied:
My update today addresses this warning.

mdavidsaver commented on modules/libcom/src/as/asLib.y:
```diff
@@ -43,6 +51,114 @@ asconfig_item:  tokenUAG uag_head uag_body
+generic_block_elem_name:  keyword
+    {
+        $$ = strdup($1);
+        if (!$$) yyerror("Out of memory");
```
Why is this branch a `strdup()`?  Is `$1` being leaked?

For that matter, could the default action be used in this case?  (as with the `keyword` rule above)

The `keyword` branch is not covered in testing.  The following triggers a crash because `$1` is apparent an integer `2` (not a float `2.0`, which I might explain.

```diff
@ -523,10 +523,11 @@ static const char supported_config_2[] = ""
 static const char supported_config_3[] = ""
     "HAG(foo) {localhost}\n"
 
     "COMPLEX_ARGUMENTS(1, WELL, \"FORMED\", ARG, LIST) {\n"
     "	ALSO_GENERIC(WELL, FORMED, ARG, LIST, 2.0) \n"
+    "	RULE(WELL, FORMED, ARG, LIST, 2.0) \n"
     "}\n"
```

```
(gdb) p yyvsp[0]
$1 = {Int64 = 4611686018427387904, Float64 = 2, 
  Str = 0x4000000000000000 <error: Cannot access memory at address 0x4000000000000000>}
```

Right, because the lexer does not yield a string for `tokenRULE`...

mdavidsaver commented on modules/libcom/src/as/asLib.y:
```diff
@@ -43,6 +51,114 @@ asconfig_item:  tokenUAG uag_head uag_body
+rule_generic_block_elem_name:  non_rule_keyword
+    {
+        $$ = strdup($1);
+        if (!$$) yyerror("Out of memory");
```
likewise.

mdavidsaver commented on modules/libcom/src/as/asLibRoutines.c:
```diff
@@ -585,27 +585,26 @@ int epicsStdCall asDumpFP(
                 pasguag = (ASGUAG *)ellNext(&pasguag->node);
                 if(pasguag) fprintf(fp,","); else fprintf(fp,")\n");
             }
-            pasghag = (ASGHAG *)ellFirst(&pasgrule->hagList);
```
A note since I had to look twice.  This assignment is redundant to an identical assignment above, and `pasghag` is not modified between the two.

mdavidsaver committed:
ascheck cleanup

mdavidsaver commented on modules/libcom/src/as/asLib.y:
```diff
@@ -43,6 +51,114 @@ asconfig_item:  tokenUAG uag_head uag_body
+    | tokenHAG
+    | tokenCALC
+    | non_rule_keyword
+    ;
+
+non_rule_keyword: tokenASG
+    | tokenRULE
+    | tokenINP
```
The `non_rule_keyword` rule is mixing token types.  The keywords `tokenASG`, `tokenRULE` do not return a value.  `tokenINP` returns an integer.  So a consumer of `non_rule_keyword` can not handle the value sensibly.  The `keyword` rule and others inherits this confusion.  Ultimately resulting in an attempt to `free()` an undefined stack value.

george-mcintyre replied:
Assigning `"ASG"`, `"RULE"`, and `"INPA"` in these cases and also in the `keyword` rule for the other cases (`"UAG"`, `"HAG"`, and `"CALC"`) effectively guarantees that the `keyword` token will always return an allocated string.

Therefore when `keyword` is used in the `generic_element` rule, and indirectly in `generic_block_list` rule it can always be freed ( `free()`) safely.   It also means that `keyword` can be handled coherently throughout the grammar without special cases.  Thx @mdavidsaver  #482fc727

mdavidsaver commented:
@george-mcintyre fyi. `antelope` is the parser generator bundled with epics-base.  It is effectively a fork of BSD yacc circa 1990 (yes really).  So it has little in the way of error detection, and will silently let you do things which more recent yacc or bison would at least warn about.

mdavidsaver committed:
asLib: fix generic lexer/parser

mdavidsaver commented on modules/libcom/src/as/asLib.y:
```diff
@@ -21,17 +22,24 @@ static ASGRULE *yyAsgRule=NULL;
 
 %start asconfig
 
-%token tokenUAG tokenHAG tokenASG tokenRULE tokenCALC
-%token <Str> tokenINP
-%token <Int> tokenINTEGER
-%token <Str> tokenSTRING
+%token <Str> tokenUAG tokenHAG tokenASG tokenRULE tokenCALC tokenINP tokenSTRING
+%token <Int64> tokenINT64
+%token <Float64> tokenFLOAT64
```
I think the token value types should be:

```yacc
%token tokenUAG tokenHAG tokenASG tokenRULE tokenCALC
%token <Str> tokenSTRING
%token <Int64> tokenINT64 tokenINP
%token <Float64> tokenFLOAT64
```

So as before `tokenUAG tokenHAG tokenASG tokenRULE tokenCALC` have no value.  Previously `tokenINP` was incorrectly marked `Str` when it actually yields an integer.

Correcting this causes `antelope` to complain:

```
"../../../../bin/linux-x86_64/antelope" -basLib  ../as/asLib.y
antelope: warning - line 77 of "../as/asLib.y", the default action assigns an undefined value to $$
antelope: warning - line 78 of "../as/asLib.y", the default action assigns an undefined value to $$
antelope: warning - line 79 of "../as/asLib.y", the default action assigns an undefined value to $$
antelope: warning - line 80 of "../as/asLib.y", the default action assigns an undefined value to $$
antelope: warning - line 81 of "../as/asLib.y", the default action assigns an undefined value to $$
antelope: warning - line 82 of "../as/asLib.y", the default action assigns an undefined value to $$
antelope: error - line 87 of "../as/asLib.y", $1 (tokenASG) is untyped
```

mdavidsaver commented:
I have rebased and made an attempt to correct the value type mixups.  With these changes `aslibtest` and `ascheck` run without crash or valgrind access errors (eg. due to attempted `strlen(2)`).

The parse error cases still leak memory, as was the case previously.  `antelope` does not know how to cleanup the parse stack on `yyerror()`.  Correcting this would be a much larger task, and not necessary for this PR.

I still do not think that there is coverage of all rules.  aka. some of the token types do not appear in ignored blocks.

mdavidsaver commented:
A couple of notes for other reviewers.

The issue I found, and hopefully fixed, has to do with special treatment of lexer keywork tokens.  eg. this:

> IGNOREME(1,WRITE) { UNKNOWN(whatever)  }

vs.

> IGNOREME(1,WRITE) { RULE(whatever)  }

For the moment, I opted to populate the `Str` value with eg. `strdup("RULE")`.  Anything more would mean a larger change to how string  values are handled on the parser stack.

Also, a note on antelope/yacc semantics which I did not understand.

```yacc
%token <Str> tokenSTRING
%type <Str> generic_block_elem_name
%type <Str> keyword

generic_block_elem_name:  keyword
                       |  tokenSTRING;
```

This rule does not leak the `Str` value.  Instead, it implicitly uses a default action `$$ = $1`.  Which is appropriate for transfering the `char*` during a reduction.

george-mcintyre committed:
Add detailed ACF forward compatibility documentation.

george-mcintyre commented:
> Core: Needs Release Notes documenting the new generic ACF syntax. Will also need adding to the RTD chapter on [Access Security](https://docs.epics-controls.org/en/latest/appdevguide/AccessSecurity.html).
 
@anjohnson addressed here #66561b84

george-mcintyre committed:
Add detailed ACF forward compatibility documentation.

george-mcintyre commented:
https://github.com/epics-docs/epics-docs/pull/154 opened to add doc to epics-docs

george-mcintyre committed:
Fix typo in PR-624 documentation

anjohnson commented on modules/libcom/src/as/asLib.y:
```diff
@@ -43,6 +61,110 @@ asconfig_item:  tokenUAG uag_head uag_body
+generic_item: tokenSTRING generic_head generic_list_block
+    {
+        yywarn("Ignoring unsupported TOP LEVEL nested block", $1);
+        free((void *)$1);
```
Don't cast arguments to `free()`, they should be pointers anyway.

george-mcintyre committed:
Simplify memory deallocation by removing redundant casts to `(void *)`.

anjohnson commented on modules/libcom/src/as/asLib.y:
```diff
@@ -175,6 +297,13 @@ rule_list_item: tokenUAG '(' rule_uag_list ')'
     }
+    | rule_generic_block_elem
+    {
+        yywarn("Ignoring RULE containing unsupported PREDICATE", $1);
```
Change warning message so non-programmers can understand what this means.

"Ignoring RULE that contains an unsupported keyword"

Similar words in lines 266, ...

george-mcintyre committed:
Clarify warnings for unsupported rule keywords in `asLib.y`.

george-mcintyre committed:
Clarify warnings for unsupported rule keywords in `asLib.y`.

anjohnson commented on modules/libcom/src/as/asLibRoutines.c:
```diff
@@ -544,7 +545,6 @@ int epicsStdCall asDumpFP(
         puag = (UAG *)ellNext(&puag->node);
     }
     phag = (HAG *)ellFirst(&pasbase->hagList);
-    if(!phag) fprintf(fp,"No HAGs\n");
```
Restore this line, or explain why it's being removed.

george-mcintyre committed:
Add check for empty HAG list in `asLibRoutines.c`.

anjohnson reviewed (changes requested):
Core Group: Looks good, we discussed these changes in today's meeting.

anjohnson commented:
Core Group: MAD will try for code coverage separately.
Ready to merge once CI passes.

mdavidsaver commented:
Merged as of e5618326c9c1d8de2461104f6f478757ed3eb1c0

mdavidsaver commented:
> Core Group: MAD will try for code coverage separately.

Of course it would be now, after merging, that I finally have a coverage report of asLib.y.  The situation is better than I expected.  In the added rules, tokenINP is not covered.  Of course, CALC and INP are not covered by the existing tests either.
